src/bin/rm.rs

```diff
@@ -1,21 +1,201 @@
+            }
+        }
+    }
 
-    if env::args().count() < 2 {
-        fail("no arguments.", &mut stderr);
+    /// Attempt to remove the file given as an input argument.
+    fn remove_file(&mut self, file: &Path) {
+        if let Err(message) = fs::remove_file(file) {
+            self.stderr.write(b"cannot remove '").try(&mut self.stderr);
+            self.stderr.write(file.to_string_lossy().as_bytes()).try(&mut self.stderr);
+            self.stderr.write(b"': ").try(&mut self.stderr);
+            print_error(message, &mut self.stderr);
+        } else if self.verbose {
+            self.stdout.write(b"removed '").try(&mut self.stderr);
+            self.stdout.write(file.to_string_lossy().as_bytes()).try(&mut self.stderr);
+            self.stdout.write(b"'\n").try(&mut self.stderr);
+            self.stdout.flush().try(&mut self.stderr);
+        }
     }
 
-    for ref path in env::args().skip(1) {
-        fs::remove_file(path).try(&mut stderr);
+    /// Attempt to remove a directory and all of it's contents if recursive mode is enabled.
+    /// If recursion is not enabled, attempt to remove the directory if it is empty.
+    fn remove_directory(&mut self, directory: &Path) {
+        // TODO: Use walkdir when it is implemented in Redox instead of fs::remove_dir_all().
+        if self.recursive {
+            if let Err(message) = fs::remove_dir_all(directory) {
+                self.stderr.write(b"cannot remove directory '").try(&mut self.stderr);
+                self.stderr.write(directory.to_string_lossy().as_bytes()).try(&mut self.stderr);
+                self.stderr.write(b"': ").try(&mut self.stderr);
+                print_error(message, &mut self.stderr);
+            } else if self.verbose {
+                self.stdout.write(b"removed directory '").try(&mut self.stderr);
+                self.stdout.write(directory.to_string_lossy().as_bytes()).try(&mut self.stderr);
+                self.stdout.write(b"'\n").try(&mut self.stderr);
+                self.stdout.flush().try(&mut self.stderr);
+            }
+        } else {
+            if let Err(message) = fs::remove_dir(directory) {
+                self.stderr.write(b"cannot remove directory '").try(&mut self.stderr);
+                self.stderr.write(directory.to_string_lossy().as_bytes()).try(&mut self.stderr);
+                self.stderr.write(b"': ").try(&mut self.stderr);
+                print_error(message, &mut self.stderr);
+            } else if self.verbose {
+                self.stdout.write(b"removed directory '").try(&mut self.stderr);
+                self.stdout.write(directory.to_string_lossy().as_bytes()).try(&mut self.stderr);
+                self.stdout.write(b"'\n").try(&mut self.stderr);
+                self.stdout.flush().try(&mut self.stderr);
+            }
+        }
     }
 }
+
+/// Print the message given by an io::Error to stderr.
+fn print_error(message: io::Error, stderr: &mut Stderr) {
```

Comment:
These programs shouldn't quit just because one argument fails out of many arguments. Instead, the errors should be handled precisely as they are. If an error occurs, that error is noted and only once all arguments have been processed will it exit with a status of 1. In other words, handling an error is more ideal than just quitting immediately.